corehq/apps/hqadmin/tasks.py

```diff
@@ -201,3 +205,84 @@ def track_pg_limits():
                 cursor.execute(f'select last_value from "{sequence}"')
                 current_value = cursor.fetchone()[0]
                 metrics_gauge('commcare.postgres.sequence.current_value', current_value, {'table': table, 'database': db})
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="4"))
+def reconcile_es_cases():
+    _reconcile_es_data('case', 'commcare.elasticsearch.stale_cases', 'reconcile_es_cases')
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="4"))
+def reconcile_es_forms():
+    _reconcile_es_data('form', 'commcare.elasticsearch.stale_forms', 'reconcile_es_forms')
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="6"))
+def count_es_forms_past_window():
+    today = date.today()
+    two_days_ago = today - timedelta(days=2)
+    four_days_ago = today - timedelta(days=4)
+    start = four_days_ago.isoformat()
+    end = two_days_ago.isoformat()
+    _reconcile_es_data(
+        'form',
+        'commcare.elasticsearch.stale_forms_past_window',
+        'es_forms_past_window',
+        start=start,
+        end=end,
+        publish=False
+    )
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="6"))
+def count_es_cases_past_window():
+    today = date.today()
+    two_days_ago = today - timedelta(days=2)
+    four_days_ago = today - timedelta(days=4)
+    start = four_days_ago.isoformat()
+    end = two_days_ago.isoformat()
+    _reconcile_es_data(
+        'case',
+        'commcare.elasticsearch.stale_cases_past_window',
+        'es_cases_past_window',
+        start=start,
+        end=end,
+        publish=False
+    )
+
+
+def _reconcile_es_data(data_type, metric, blob_parent_id, start=None, end=None, republish=True):
+    if not start:
+        today = date.today()
+        two_days_ago = today - timedelta(days=2)
```

Comment:
Continuing the thread on date boundaries from above (now that it's kind of in the wrong place). My remaining question is whether we're reprocessing data here all the way up until the present moment. Does `end` in `stale_data_in_es` refer to the midnight preceding that day or the midnight after that day? I would think it means up to the midnight preceding that day, and so wouldn't include any changes made _on_ that day.

If it's the other way around, that means this reprocesses data all the way up to the present moment. This opens it up to what I think would be a common occurrence where this process would confuse the normal delay in the SQL to ES pipeline with docs being stale.

Reply:
Okay I offlined with Cal. I had missed what he was saying almost entirely. The issue is that we're operating off of the modification date of the case, not the timestamp of each individual change to the case, so a case that's always updated at 1am but always stale will continuously fall into that gap period. Makes sense